snorkel/labeling/model/tuner.py

```diff
@@ -0,0 +1,157 @@
+import json
+import os
+from typing import Any, Dict, Optional, Tuple
+
+import numpy as np
+import ray
+from ray.tune import Experiment, Trainable
+
+from snorkel.labeling import LabelModel
+
+
+class LabelModelTuner(object):
+    """A hyperparameter tuner for the LabelModel using Ray Tune (https://ray.readthedocs.io/en/latest/tune.html).
+
+    Example
+    -------
+    >>> L_train = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+    >>> L_dev = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+    >>> Y_dev = np.array([1, 1, 0, 1])
+    >>> label_model = LabelModel(cardinality=2)
+    >>> label_model_tuner = LabelModelTuner(label_model, L_train, L_dev, Y_dev)
+    >>> label_model_tuner = LabelModelTuner(label_model, L_train, L_dev, Y_dev, fit_config={"class_balance" : np.array([0.5, 0.5])})
+
+    Parameters
+    ----------
+    label_model
+        LabelModel instance
+    L_train
+        A matrix with values in {-1,0,1,...,k-1}
+    L_dev
+        A matrix with values in {-1,0,1,...,k-1}
+    Y_dev
+        Gold labels for dev set for tuning hyperparameters
+    metric
+        Metric to maximize while hyperparameter tuning
+    fit_config
+        Arguments for changing LabelModel fit() config defaults
+    """
+
+    def __init__(
+        self,
+        label_model,
+        L_train: np.ndarray,
+        L_dev: np.ndarray,
+        Y_dev: np.ndarray,
+        metric: str = "accuracy",
+        fit_config: Optional[Dict[str, Any]] = {},
+    ) -> None:
+
+        self.label_model = label_model
+        self.L_train = L_train
+        self.L_dev = L_dev
+        self.Y_dev = Y_dev
+        self.metric = metric
+        self.fit_config = fit_config
+
+    def _trainable(self):  # type: ignore
+        """Return a Trainable class definition for Ray tuner with some user-defined parameters from LabelModelTuner class."""
+
+        class LabelModelTrainable(Trainable):
+            """Trainable class for LabelModel for Ray tuner."""
+
+            def _setup(self, config: Optional[Dict[str, Any]]):
+                self.tuner = LabelModelTrainable.tuner  # type: ignore
```

Comment:
I'm guessing this is some ray pattern, but i don't see why this is necessary

Reply:
Ray requires a Trainable with specific forms for the _train(), _setup() methods, but doesn't have a clean way of taking in additional args for those methods. This links to the overall LabelModelTuner class when Trainable is defined so it has access to the args that are constant across all runs and required across fit/score etc.